### ISSUE TITLE
Failed assertion: boolean expression must not be null

### ISSUE THREAD

**Issue by weiliang0626** [lukef/qr.flutter#51]
**Describe the bug**
I use
 01$01$51010002271907195487$豫A12348,黄色,川E0697挂,黄色$张一,110101******011111,张二,110101******210011$10.0$1971,罐装,10.0,吨$OVprCuw5UXYq7PK8dc/2MM3htIwaxQOu04BBU2JAcca8aKE4lqNbsm5j7hRJKKpiZY3P7A9g==
to creat a qr,but reported Failed assertion: boolean expression must not be null,
it break at
if (_qr.isDark(y, x)) {
          final Rect squareRect = Rect.fromLTWH(x * squareSize, y * squareSize,
              squareSize + pxAdjustValue, squareSize + pxAdjustValue);
          canvas.drawRect(squareRect, _paint);
        }

**Comment by lukef** [comment]
Can you post your code for the `QrImage`? I tried it using my current build and it seems to work.

**Comment by weiliang0626** [comment]
this is my code :
QrImage(
                  data:'01/01/51010002271907197334/豫A12348,黄色,川E0599挂,黄色/张一,110101**101******210011/10.0/1971,罐装,10.0,吨/p/QytdcxxXEVxW9DtWzoY0KhBButCeqqoK8lIpfr1ul3vLv4CrAsqNl0teAVDqvNy/M8FTJRFxmv5CBxPiFzrA==',
                  size: 180,
                  foregroundColor: AppColor.color51,
                ),
my version is 2.1.0

**Comment by lukef** [comment]
I haven't been able to reproduce this on `2.1.0` or on the `3.0.0-dev` version. What version of Flutter are you using?

**Comment by JotaFerreira** [comment]
Same problem.

Channel beta, v1.7.8+hotfix.3

**Comment by weiliang0626** [comment]
ma current flutter version is Flutter 1.5.4-hotfix.2 • channel ,
Tools • Dart 2.3.0 (build 2.3.0-dev.0.5 a1668566e5)；
when I upgrade to Flutter 1.7.8+hotfix.3 • channel stable
Tools • Dart 2.4.0,
Same thing

**Comment by avinath1998** [comment]
Yep, i'm getting the same bug as well.
Channel stable, v1.7.8+hotfix.3

**Comment by lukepighetti** [comment]
With this pubspec.yaml

```
name: sandbox
description: A new Flutter project.

dependencies:
  flutter:
    sdk: flutter
  qr_flutter: ^2.1.0+55

dev_dependencies:
  flutter_test:
    sdk: flutter

flutter:
  uses-material-design: true

```

And this minimal reproduction

```
import 'package:flutter/material.dart';
import 'package:qr_flutter/qr_flutter.dart';

main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Center(
          child: QrImage(
            data:
                '01/01/51010002271907197334/豫A12348,黄色,川E0599挂,黄色/张一,110101**101******210011/10.0/1971,罐装,10.0,吨/p/QytdcxxXEVxW9DtWzoY0KhBButCeqqoK8lIpfr1ul3vLv4CrAsqNl0teAVDqvNy/M8FTJRFxmv5CBxPiFzrA==',
            size: 180,
            foregroundColor: AppColor.color51,
          ),
        ),
      ),
    );
  }
}

class AppColor {
  static final color51 = Colors.blue;
}
```

And this version of Flutter

```
$ flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel dev, v1.8.2, on Mac OS X 10.14.5 18F132, locale en-US)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 10.3)
[✓] iOS tools - develop for iOS devices
[✓] Android Studio (version 3.1)
[✓] VS Code (version 1.36.1)
[✓] Connected device (2 available)

• No issues found!
```

This is drawn 

![IMG_E76D93DCFA35-1](https://user-images.githubusercontent.com/666539/61818786-93e52d00-ae1f-11e9-9373-6d8225d968e7.jpeg)


And this error is thrown

```
flutter: ══╡ EXCEPTION CAUGHT BY RENDERING LIBRARY ╞═════════════════════════════════════════════════════════
flutter: The following assertion was thrown during paint():
flutter: Failed assertion: boolean expression must not be null
flutter:
flutter: Either the assertion indicates an error in the framework itself, or we should provide substantially
flutter: more information in this error message to help you determine and fix the underlying cause.
flutter: In either case, please report this assertion by filing a bug on GitHub:
flutter:   https://github.com/flutter/flutter/issues/new?template=BUG.md
flutter:
flutter: User-created ancestor of the error-causing widget was:
flutter:   Padding
flutter:   file:///Users/lukepighetti/flutter/.pub-cache/hosted/pub.dartlang.org/qr_flutter-2.1.0+55/lib/src/qr_image.dart:49:18
flutter:
flutter: When the exception was thrown, this was the stack:
flutter: #0      QrPainter.paint  (package:qr_flutter/src/qr_painter.dart:71:17)
flutter: #1      RenderCustomPaint._paintWithPainter 
package:flutter/…/rendering/custom_paint.dart:528
flutter: #2      RenderCustomPaint.paint 
package:flutter/…/rendering/custom_paint.dart:566
flutter: #3      RenderObject._paintWithContext 
package:flutter/…/rendering/object.dart:2091
flutter: #4      PaintingContext.paintChild 
package:flutter/…/rendering/object.dart:172
flutter: #5      RenderShiftedBox.paint 
package:flutter/…/rendering/shifted_box.dart:70
flutter: #6      RenderObject._paintWithContext 
package:flutter/…/rendering/object.dart:2091
.....
flutter: (elided 8 frames from package dart:async and package dart:async-patch)
flutter:
flutter: The following RenderObject was being processed when the exception was fired: RenderCustomPaint#76d3b:
flutter:   creator: CustomPaint ← Padding ← ConstrainedBox ← Container ← LayoutBuilder ← QrImage ← Center ←
flutter:     MediaQuery ← LayoutId-[<_ScaffoldSlot.body>] ← CustomMultiChildLayout ← AnimatedBuilder ←
flutter:     DefaultTextStyle ← ⋯
flutter:   parentData: offset=Offset(10.0, 10.0) (can use size)
flutter:   constraints: BoxConstraints(w=160.0, h=160.0)
flutter:   size: Size(160.0, 160.0)
flutter: This RenderObject has no descendants.
flutter: ════════════════════════════════════════════════════════════════════════════════════════════════════
```

**Comment by lukepighetti** [comment]
As a note, the non-roman characters do not affect the outcome.

**Comment by lukepighetti** [comment]
> this is my code :
> 
> ```
> QrImage(
>   data:'01/01/51010002271907197334/豫A12348,黄色,川E0599挂,黄色/张一,110101**101******210011/10.0/1971,罐装,10.0,吨/p/QytdcxxXEVxW9DtWzoY0KhBButCeqqoK8lIpfr1ul3vLv4CrAsqNl0teAVDqvNy/M8FTJRFxmv5CBxPiFzrA==',
>   size: 180,
>   foregroundColor: AppColor.color51,
> ),
> ```
> 
> my version is `2.1.0`.

Hi there, if you change your Qr version to 9 it should work.


```dart
QrImage(
  data:
      '01/01/51010002271907197334/豫A12348,黄色,川E0599挂,黄色/张一,110101**101******210011/10.0/1971,罐装,10.0,吨/p/QytdcxxXEVxW9DtWzoY0KhBButCeqqoK8lIpfr1ul3vLv4CrAsqNl0teAVDqvNy/M8FTJRFxmv5CBxPiFzrA==',
  size: 180,
  foregroundColor: AppColor.color51,
  version: 9,
),
```

**Comment by lukef** [comment]
If you set a version (e.g.: `version: 6`) do you still see the issue? The next version `3.0.0` has an `auto` setting and will handle version conflicts more gracefully. I suspect the issue is just that the string is too long for the default version (`4`) and the error is just not helpful.

**Comment by lukef** [comment]
I'll leave this issue open until `3.0.0` lands (this week)

**Comment by weiliang0626** [comment]
@lukepighetti  @lukef  I set a version 9,it works,thanks alot.

**Comment by lucadelc** [comment]
So it will be fixed with the "3.0.0" version?

**Comment by lukef** [comment]
The error was an unfortunate uncaught exception (which is resolved in `3.0.0`) but in general if you use a version that wont support the content you're trying to capture then the library will throw an exception. This specific use case in `3.0.0` will handle the error gracefully. `3.0.0` also introduces an `auto` version that will figure out the appropriate minimum version for you (see #3 for details).

**Comment by lukef** [comment]
Fixed in 3.0. It should be on pub soon.
